### PR TITLE
CRM-19813 - Unify Symfony Events and hooks

### PR DESCRIPTION
**Pull request by totten**
Civi has three notions of events:

 * __Hooks__: These are available to extensions (either native or CMS-based). They are usually described by a function signature. Hooks are the most recognizable/supported events, but they also have limitations (hard to inspect, prioritize, cancel, etc) because they must meet the lowest-common-denominator of CMS features.
 * __Symfony Events__: These are generally only used within core. They have more advanced features but haven't been presented for downstream usage. There's [documentation](http://symfony.com/doc/current/components/event_dispatcher.html) available from Symfony.
 * __Component-actions__: These are adhoc functions which loop through core's list of "components". These are generally not well-understood.

This PR aims to unify two of them -- hooks and Symfony Events. More specifically:

 * Symfony Event Dispatcher become the **canonical event system**. All events are Symfony events. You can register event subscribers in core (or elsewhere).
 * Hooks are a **special case** in which (a) the event name looks like `hook_civicrm_foo` and (b) the event object extends `GenericHookEvent`. In this special case, the event will be simultaneously emitted through various CMS event systems (Drupal Hooks, WP Actions, etc).

If you were previously unaware that Symfony Events existed in core, then it seem like "yet another way". It is actually a unification of two existing ways. To understand how this unification positions us, consider an analogy to APIv3 CRUD:

 * For APIv3 CRUD, the public API has one set of *semantics* and multiple *syntaxes*. The semantics are based on entities+actions+params. The *syntax* for calling APIv3 varies depending on the technique/context (e.g. PHP logic vs PHP test vs JS vs REST vs CLI). There is a canonical mechanism which supports all features (`civicrm_api(...)`) and can be used in most PHP contexts, but some variants (like Smarty or CLI) can't easily support all features.
 * For hooks/events, the public API has one set of *semantics* and multiple *syntaxes*. The semantics are based on hook-name+params. The *syntax* for listening to an event varies depending on the technique/context (e.g. core code vs Civi extension vs Drupal module vs WordPress plugin vs Joomla plugin). There is a canonical mechanism which supports all features (`Civi::service('dispatcher')`) and can be used in most PHP contexts, but some variants (like Drupal/Joomla/WordPress modules) can't easily support all features.

Additional notes:

 * This means that core code can listen to hooks (the same way it already listens to events).
 * Hooks are the *most visible*/*most accessible* type of events.  However, it's possible for any extension to work with the full features of the event-dispatcher.
 * Use `cv debug:event-dispatcher` to inspect the list of event listeners. 
 * Some discussion of the design of `GenericHookEvent` is included in the relevant commit ("GenericHookEvent - Bridge between Symfony Events and hooks").
 * This is a WIP. There's another hook that I'm working on (for management of dynamic-assets) that is part of the impetus.
 * This is a WIP. To get full support for all hooks in Symfony Events, we need small patches to each stub in `CRM_Utils_Hook`. I've adapted ~1/3 so far.

---

 * [CRM-19813: Hook priorities and core hooks to support LExIM](https://issues.civicrm.org/jira/browse/CRM-19813)

### CODE REVIEW

**Comment by nganivet** [comment]
Also ... what are the documentation changes required for this? Do we need to add a section with the additional events now available to extensions?

**Comment by totten** [comment]
I've removed the WIP flag because I've addressed the only blocker.

@nganivet I think the most helpful thing would be to smoke-test the patch in a few use-cases:
  * **What**: Apply the patch. Take a Drupal module, WordPress plugin, or Civi extension that you know well and which relies on core hooks. (Ex: IIRC, Sparkpost relies on several hooks.) Try using it like a regular person and see if it works. This doesn't need to be thorough evaluation of every codepath -- just a basic test (like you might do in a demo to a new user).
  * **Why**: Although I think this PR is pretty decent in terms of refactoring technique and unit-tests, there are still gaps in the test-coverage, and it would be pretty embarrassing if it broke the entire hook system in some common scenario. We're not likely to find anyone who can analytically prove the patch's safety, but a few light/diverse checks would helpful (per [Swiss Cheese Model](https://en.wikipedia.org/wiki/Swiss_cheese_model)). 

In my ideal world, our stack of cheese slices would include somewhere within it:
 * A Drupal module, a WordPress plugin, a Joomla plugin, a Civi extension
 * A hook that reads data -- and a hook that modifies data
 * A hook that fires in a normal web-based UI and a hook that fires in backend/web-service context

If a couple people could try their own slices (whatever's familiar/easy for you) and report back, then I'll try a slice to fill in whatever gap remains.

**Comment by nganivet** [comment]
@totten I am putting testing this on my todo list, but am traveling starting tomorrow until April 4th. Will see if I have time while on the road, but most realistically not.

**Comment by totten** [comment]
Oop, found a problem. It arises when a hook's has an input `$name` like:

```php
  public static function contactListQuery(&$query, $name, $context, $id) {
```

If someone tries to use a `GenericHookEvent` to access the corresponding `$event->name`, it leads to a conflict because the base class `Event` has its own private property named `name`.

A simple workaround is to change `$name` to something more precise. Fortunately, this should have no impact on existing listeners.

File: CRM/Utils/Hook.php [hunk 1]
```diff
@@ -298,10 +340,8 @@ public function requireCiviModules(&$moduleList) {
    */
   public static function pre($op, $objectName, $id, &$params) {
     $event = new \Civi\Core\Event\PreEvent($op, $objectName, $id, $params);
-    \Civi::service('dispatcher')->dispatch("hook_civicrm_pre", $event);
-    \Civi::service('dispatcher')->dispatch("hook_civicrm_pre::$objectName", $event);
```
Comment (eileenmcnaughton) [inline CRM/Utils/Hook.php]: what does this change do?

File: CRM/Utils/Hook.php [hunk 2]
```diff
@@ -122,13 +123,54 @@ public function __construct() {
...
+      $event = \Civi\Core\Event\GenericHookEvent::createOrdered(
+        $names,
+        array(&$arg1, &$arg2, &$arg3, &$arg4, &$arg5, &$arg6)
+      );
+      \Civi::dispatcher()->dispatch('hook_' . $fnSuffix, $event);
+      return $event->getReturnValues();
+    }
+    else {
+      $count = is_array($names) ? count($names) : $names;
+      return $this->invokeViaUF($count, $arg1, $arg2, $arg3, $arg4, $arg5, $arg6, $fnSuffix);
```
Comment (eileenmcnaughton) [inline CRM/Utils/Hook.php]: this is the legacy support

**Comment by eileenmcnaughton** [comment]
I've taken a look at this and feel it's OK to merge provided it gets some more testing in the rest of the release process. We are a good time in the cycle now. We do have reasonable in-core testing.

 I'll test against the civixero extension which implements it's own hook extension, @seamuslee001 can you check on your multisite.

**Comment by eileenmcnaughton** [comment]
Note we should also notifiy the dev list

File: CRM/Utils/Hook.php (hunk 1 above)
Comment (totten) [inline CRM/Utils/Hook.php]: Good question. This change come from c73e3098.

The old lines were part of a quick-and-dirty attempt to support hook-events in core. (You might refer to them as dual-emit because they were emitted once through Symfony `EventDispatcher` and once through CMS.) However, that approach required a bunch of boilerplate: you write a new PHP file (`Civi/Core/Event/PreEvent.php`) for *every hook*, and then patch *every hook* to instantiate+dispatch it.

In the revised code, lines 333-335 are redundant. If you call `dispatch('hook_foo', ...)`, then the event is dispatched everywhere. You don't need extra classes because all hooks can use `GenericHookEvent`.

There were 3 or 4 hooks which had the old boilerplate classes. I took an approach which felt more conservative -- the boilerplate classes are still there, but they now extend `GenericHookEvent`.

In retrospect, it would have been fair-game to entirely remove those boilerplate classes. Grepping universe, it appears that `PreEvent` (etal) have been internal to `civicrm-core`. But I don't mind the above because it's technically better refactoring technique.

**Comment by eileenmcnaughton** [comment]
@nganivet can you please continue to keep doing testing on your todo list - albeit against master now

**Comment by christianwach** [comment]
@totten (Reposting here in case you miss my message on MM) On further testing, I notice that WordPress does not receive callbacks from `hook_civicrm_entityTypes` but I'm unable to trace why it's failing.

**Comment by christianwach** [comment]
@totten I presume from 4d8e83b6770f500adb949524f1e90d0f4b4e58f6 that `hook_civicrm_entityTypes` needs special handling. I can see the successful callback in the output in my [modified Show All The Hooks extension](https://github.com/christianwach/contrib.showallthehooks/tree/wp-actions) but only for the extension.

**Comment by christianwach** [comment]
@totten Here's what's going on: 

* When the code declaring the callback is in a Civi extension, the `add_action()` call declared there occurs *after* `hook_civicrm_entityTypes` has already fired. This accounts for the lack of activity in the callback.

* When I declare a callback in an ordinary WordPress plugin, `hook_civicrm_entityTypes` *is* triggered and the callback receives data during a standard page load.

In summary - CiviCRM extensions cannot inspect `$entityTypes` via the `add_action()` method - they *must* use the Drupalesque function-name-as-callback method. Vanilla WordPress plugins can use `add_action()`.

Note: I have only tested this with standard page loads - not via cron or any other entry route.

**Comment by totten** [comment]
Thanks again @christianwach. For anyone who stumbles across this, we chatted more over Mattermost, and I think that phenomenon makes sense -- you should pick the notation that is canonical for your package type. (For WP plugin, use WP notation. For Drupal module, use Drupal notation.)
